monitoring/nethealth.go

```diff
@@ -0,0 +1,483 @@
+	return &netStats{
+		TTLMap:             holster.NewTTLMap(mapCapacity),
+		packetLossCapacity: packetLossCapacity,
+	}
+}
+
+// Get returns the peerData for the specified peer.
+func (r *netStats) Get(peer string) (data peerData, err error) {
+	r.Lock()
+	defer r.Unlock()
+
+	value, ok := r.TTLMap.Get(peer)
+	if !ok {
+		return peerData{packetLoss: make([]float64, 0, r.packetLossCapacity)}, nil
+	}
+
+	data, ok = value.(peerData)
+	if !ok {
+		return data, trace.BadParameter("expected %T, got %T", data, value)
+	}
+
+	return data, nil
+}
+
+// Set maps the specified peer and data.
+func (r *netStats) Set(peer string, data peerData) error {
+	r.Lock()
+	defer r.Unlock()
+	return r.TTLMap.Set(peer, data, netStatsTTLSeconds)
+}
+
+// peerData keeps trock of relevant nethealth data for a node.
+type peerData struct {
+	// prevRequestTotal keeps track of the previously recorded total number of
+	// requests sent to the peer. This is necessary to calculate the number of
+	// new requests since the last check.
+	prevRequestTotal float64
+	// prevTimeoutTotal keeps track of the previously recorded total number of
+	// requests sent to the peer and timed out. This is necessary to calculate
+	// the number of new timeouts since the last check.
+	prevTimeoutTotal float64
+	// packetLoss records the packet loss in percentages between check intervals.
+	packetLoss []float64
+}
+
+// networkData contains a request and timeout counter.
+type networkData struct {
+	// requestTotal specifies the total number of requests sent.
+	requestTotal float64
+	// timeoutTotal specifies the total number of requests that timed out.
+	timeoutTotal float64
+}
+
+const (
+	nethealthCheckerID = "nethealth-checker"
+	// peerLabel specifies the label to collect peer name from nethealth metrics.
+	peerLabel = "peer_name"
+	// nethealthNamespace specifies the k8s namespace that nethealth exists within.
+	nethealthNamespace = "monitoring"
+	// echoRequestLabel defines the metric family label for the echo request counter.
+	echoRequestLabel = "nethealth_echo_request_total"
+	// echoTimeoutLabel defines the metric family label for the echo timeout counter.
+	echoTimeoutLabel = "nethealth_echo_timeout_total"
+
+	// defaultNethealthPort defines the default nethealth port.
+	defaultNethealthPort = 9801
+
+	// defaultNetStatsInterval defines the default interval duration for the netStats.
+	defaultNetStatsInterval = 5 * time.Minute
+
+	// netStatsCapacity sets the number of TTLMaps that can be stored.
+	// This will be the size of the cluster -1.
+	netStatsCapacity = 1000
+
+	// netStatsTTLSeconds defines the time to live in seconds for the stored
+	// netStats. This ensures the checker does not hold on to unused data when
+	// a member leaves the cluster.
+	netStatsTTLSeconds = 60 * 60 * 24 // 1 day
+
+	// packetLossThreshold defines the packet loss percentage used to determine
+	// if overlay network communication with a peer is unhealthy.
```

Comment:
Needs adding that the loss needs to be consistently observed over the entire interval and not on individual samples